### ISSUE TITLE
Review top-level docs for prospective kernel authors

### ISSUE THREAD

**Issue by minrk** [jupyter/jupyter#216]
We should take a pass on the top-level Jupyter docs with prospective kernel authors in mind.

Key information:

- Not necessarily familiar with Jupyter and/or Python
- only assume they arrive with the keyword "kernel" to work with, and the knowledge that they might want to make one
- might also have "protocol" or "message spec" 

Should be able to get some idea about:

- why would I want to write a kernel?
- how much work is it?
- what information do I need to get started (protocols, etc.)?

They should probably end up here: http://jupyter-client.readthedocs.io/en/latest/kernels.html and there's probably more information to add once they arrive.

/cc @willingc @williamstein

Also cc @captainsafia who I believe has some good materials about writing new Jupyter kernels.

**Comment by captainsafia** [comment]
My [PyData Chicago](https://www.youtube.com/watch?v=LeMwtpYootY) discussed a bit about building kernels and the general JMP infrastructure. I also at some point had a play-along tutorial-style blog post running about building a kernel from scratch but I'll have to dig up where that went.

**Comment by jasongrout** [comment]
It would also be helpful to point to an implementation of a very simple kernel (perhaps just echoing execute requests), a pexpect-based kernel, and the wiki list of kernels.

**Comment by takluyver** [comment]
The docs on writing wrapper kernels [do include](http://jupyter-client.readthedocs.io/en/latest/wrapperkernels.html#example) an example of a kernel echoing execute requests, though this doesn't currently include packaging and distribution, just the code itself.

Doug Blank's [simple_kernel](https://github.com/dsblank/simple_kernel) is meant to illustrate what's needed to write a compliant kernel from scratch (not reusing the `jupyter_client` machinery).

bash_kernel was also originally intended as a simple example of a pexpect-based kernel, but it's grown quite a bit as people wanted to do more and more with it.

**Comment by jasongrout** [comment]
> The docs on writing wrapper kernels do include an example of a kernel echoing execute requests,

Thanks. I thought there was something like this, but somehow missed this [second page](http://jupyter-client.readthedocs.io/en/latest/wrapperkernels.html) when I read through the first page of the jupyter_client docs (there is a prominent link, but after reading through the first page, I started clicking on the detail links about the wire protocol, etc.). And I see this second page also links to the bash pexpect kernel,

**Comment by dsblank** [comment]
You might also consider linking to https://github.com/Calysto/metakernel because the goal of metakernel is to provide common infrastructure (magics, shell, parallel execution, data sharing, etc.) for other kernels. When I first started on a C#-based shell, I thought I had to write the kernel in C#. If I had to do it over again, I would not have written that. Too much work for too little. It would be nice to have a section discussing how one would decide if one should write from scratch (in host language or Python), or use a common core (like metakernel).

**Comment by minrk** [comment]
Thanks, everyone! Perhaps we just need to make sure that people starting [here](https://jupyter.readthedocs.io) can find this great stuff. It might be worth a dedicated page for kernel authors, expanding beyond the few bullets we currently have under Kernels.

**Comment by willingc** [comment]
How about adding a section under Architecture Guides for "Programming Language Kernels" or a top-level item for "Programming Language Kernels"?

"Programming Language Kernels"
- Brief section of language agnostic value of Jupyter
- Language kernels available
   * Maintained by Project Jupyter
   * Maintained by Community
- How to create a new language kernel -> Link to jupyter_client docs
   * Writing a new kernel -> Link to jupyter_client and @captainsafia info
   * Using metakernel as a template

**Comment by minrk** [comment]
@willingc that sounds like a great start.

**Comment by williamstein** [comment]
When looking at http://jupyter-client.readthedocs.io/en/latest/wrapperkernels.html#example, I think:

1. Is this the *minimal* possible example?  In particular, are all four of 'status', 'execution_count', 'payload', and 'user_expressions" necessary?

2. It would be great to see a minimal example in some language other than Python, which explicitly exposes any implicit complexity hidden in `from ipykernel.kernelbase import Kernel`.  One of the most valuable things about jupyter kernels is that some of them are not written in Python.  I was looking at the R jupyter kernel recently, and it seems to be several thousand lines of code, so these things can get complicated!


Incidentally, when I clicked on http://jupyter-client.readthedocs.io/en/latest/messaging.html
I saw "Our current implementation uses JSON explicitly as its message format, but this shouldn’t be considered a permanent feature. As we’ve discovered that JSON has non-trivial performance issues due to excessive copying, we may in the future move to a pure pickle-based raw message format."  I assume that now that Jupyter is trying to be language agnostic this should be rewritten to suggest using Protocol Buffers rather than "pure pickle"; protobufs are pretty much the canonical language agnostic solution to this sort of problem...

**Comment by jasongrout** [comment]
@williamstein, @takluyver's answer above helps with your 'minimal example' question:

> Doug Blank's [simple_kernel](https://github.com/dsblank/simple_kernel) is meant to illustrate what's needed to write a compliant kernel from scratch (not reusing the jupyter_client machinery).

**Comment by takluyver** [comment]
> are all four of 'status', 'execution_count', 'payload', and 'user_expressions" necessary?

It's a bit vague whether empty fields can be omitted; the docs don't explicitly say that's OK, but in practice it usually will be. I think it's better to include those fields so that the example is inarguably compliant with the protocol.

> the R jupyter kernel recently, and it seems to be several thousand lines of code

I think that only a few hundred of that is actually handling the messaging protocol. Like bash_kernel, it has grown gradually more complex over time as various little affordances have been added to support stuff users want to do in it.

> I assume that now that Jupyter is trying to be language agnostic this should be rewritten to suggest using Protocol Buffers rather than "pure pickle"

TBH I'd probably just get rid of that sentence altogether. It was written when we expected the ecosystem to be multiple frontends to one Python kernel, whereas we now have much greater diversity of kernels. Performance hasn't really been a big issue (except maybe for ipyparallel, which is rather its own thing), and we haven't talked at all about switching away from JSON for years.

**Comment by minrk** [comment]
Yeah, it's really only in IPython parallel that alternate serializers are used, and I mostly use msgpack there. We should remove this from the general messaging doc.

**Comment by takluyver** [comment]
See jupyter/jupyter_client#222
